### ISSUE TITLE
subscribe_unsubscribe's call to self.command_queue.get(0) can return undefined.

### ISSUE THREAD

**Issue by davidmurdoch** [NodeRedis/node_redis#1230]
* **node_redis Version**: 2.7.1
* **Redis Version**: 3.2.4
* **Platform**: Node.js 6.10.2 on Heroku
* **Description**: `subscribe_unsubscribe`'s call to [`self.command_queue.get(0)`](https://github.com/NodeRedis/node_redis/blob/master/index.js#L730) can return `undefined`. This will cause the [`command_obj.args`](https://github.com/NodeRedis/node_redis/blob/master/index.js#L747) check to throw.

I'm unsure why this happens on our server on occasion, and I have been unable to reliably reproduce this issue. The only thing I can think of that might be causing the problem, based on the comments in the `subscribe_unsubscribe` function about the optional callback, is that we omit the callback in our `subscribe` and `unsubscribe` calls, e.g., `pubSubClient.subscribe([channel1, channel2, etc...]);`.

Let me know if you need any more information or if there is anything I can do to help.

Example stack trace:
```
/app/node_modules/redis/index.js:747 
      if (command_obj.args.length === 1 || self.sub_commands_left === 1 || command_obj.args.length === 0 && (count === 0 || channel === null)) { 
              ^ 
  TypeError: Cannot read property 'args' of undefined 
      at subscribe_unsubscribe (/app/node_modules/redis/index.js:747:20) 
      at return_pub_sub (/app/node_modules/redis/index.js:796:9) 
      at RedisClient.return_reply (/app/node_modules/redis/index.js:828:9) 
      at JavascriptRedisParser.returnReply (/app/node_modules/redis/index.js:192:18) 
      at JavascriptRedisParser.execute (/app/node_modules/redis-parser/lib/parser.js:574:12) 
      at Socket.<anonymous> (/app/node_modules/redis/index.js:274:27) 
      at emitOne (events.js:96:13) 
      at Socket.emit (events.js:188:7) 
      at readableAddChunk (_stream_readable.js:176:18) 
      at Socket.Readable.push (_stream_readable.js:134:10) 
```

**Comment by BridgeAR** [comment]
Hey @davidmurdoch this does indeed look very wrong.

Omitting the callback should not be an issue though. Without knowing how your code looks like, it's difficult to debug this though. Do you have some code snippet for me? Or at least all used commands from that client including there likely execution order?

**Comment by davidmurdoch** [comment]
I'll throw some logging in for all redis calls and see if we can get the issue to reoccur (it might be a few days or weeks, as it is pretty infrequent).

**Comment by davidmurdoch** [comment]
I just glanced at all occurances of this error and it seems to be happening when we subscribe to a bunch of channels and publish to some of these same channels at nearly the same exact moment.

so:
```
pubSubClient.subscribe([channel1, channel2, channel3]); // don't wait for async op to finish...
doSomethingAsync(function callback(err, channels, data) {
    // where channels contains many of the same channels we just subscribed to.
    channels.forEach(function (channel) {
        pubSubClient.publish(channel, data);
    });
});
```

the issue might be related to the fact that we are in the middle of subscribing to a channel that we are also publishing to all on the same node process.

I'm still going to add the logs and let you know if that turns up any more clues.

**Comment by davidmurdoch** [comment]
The logs helped track down the issue.

Our subscribe logic is something like this:

```
var channels = ["channel1", "channel2", ...];
pubSub.subscribe(channels);
channels.unshift("key"); // <---- the problem
client.multi().sadd(channels).exec();
```

Which both subscribes and sadd's successfully. However, because we modified the array passed to `subscribe`, and `subscribe` hangs on to this array (it doesn't copy it), when it enters `subscribe_unsubscribe`   and counts down the number of replies remaining for this command it never gets to `0` because the array is 1 greater than it expects.

Because of this, sometimes the `command_queue` is emptied prematurely due to this mismatch and it explodes with the above error.

The easy fix for me now is to copy my `channels` array before prepending to it, though I'd rather not. :-)

If `node_redis` wants to fix this I think adding a `len` property to the `Command` class (simply `this.len = args.length`) and then checking that `len` property in `subscribe_unsubscribe` instead of `args.length` would be work well.

I can submit a pull request if you'd like. Just let me know!

**Comment by BridgeAR** [comment]
@davidmurdoch thanks a lot for looking deeper into this and great work!

**Comment by BridgeAR** [comment]
Fixed in 2.8.0
